### PR TITLE
Query: Translate GroupBy-Aggregate with condition

### LINKED ISSUE #11711: Query: GroupBy Where Aggregate translation to server
Currently if there is predicate after grouping and before applying Aggregate operation we block the translation to server and do streaming GroupBy. Aggregate operator argument can take Case Block to do conditional aggregate and allow us to generate Group By clause in SQL.
Structure would be
```SQL
COUNT(
    CASE 
        WHEN predicate 
        THEN projection
        ELSE NULL
    END
)
```

### LINKED ISSUE #18836: Can't translate GroupBy with  Conditional Aggregate
this is my Class and code to 
``` C#
public enum StatusType
{
  Waiting = 0,
  Reading = 1,
  Printing = 2
}
public class DistributeListDetail
{
 public int Id {get;set;}
 public string AddressCode {get;set;}
 public StatusType Status {get;set;}
}
```

now i have a Query like this 
``` C#
var  result = Context.DistributeListDetails.Join().Where()
.GroupBy(x=> new { Path = x.AddressCode.SubString(0,4) })
.Select (x=> new {
 AddressCodeGrouped = x.Key.Path,                    
                    DistributedCustomerCount = x.Count(),
                    WaitingCount = x.Count(c => c.Status == StatusType.Waiting),
                    ReadCount = x.Count(c => c.Status == StatusType.Reading ),
}).ToListAsync();
```
the DistributedCustomerCount , WaitingCount  , ReadCount  is always Equals . maybe the 
Conditionals at Count method not Translating.

EF Core version: 3.0
Database provider: (Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (.NET Core 3.0)
Operating system: Windows 10
IDE: (Visual Studio 2019 16.3)

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #18836
Resolves #11711

### CODE REVIEW

File: test/EFCore.Specification.Tests/Query/NorthwindGroupByQueryTestBase.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Linq;
 using System.Threading.Tasks;
+using Castle.Components.DictionaryAdapter;
```
Comment (maumar) [inline test/EFCore.Specification.Tests/Query/NorthwindGroupByQueryTestBase.cs:7]: remove

Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/NorthwindGroupByQueryTestBase.cs:7]: Will remove in next PR.

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -498,29 +457,160 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
-                    throw new InvalidOperationException(
-                        TranslationErrorDetails == null
-                            ? CoreStrings.TranslationFailed(methodCallExpression.Print())
-                            : CoreStrings.TranslationFailedWithDetails(methodCallExpression.Print(), TranslationErrorDetails));
-                }
+                    Expression result;
+                    switch (methodCallExpression.Method.Name)
+                    {
+                        case nameof(Enumerable.Average):
+                            if (methodCallExpression.Arguments.Count == 2)
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:469]: there is a predicate pattern and select pattern - could DRY those, but maybe not worth

Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:469]: Not worth or easy.
Apply selector never returns null but apply predicate can.

**Pull request merged**